File: ODK.Website.Angular/src/app/services/chapters/chapter-admin.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable, of } from 'rxjs';
import { map } from 'rxjs/operators';

import { catchApiError } from '../http/catchApiError';
import { Chapter } from 'src/app/core/chapters/chapter';
import { ChapterAdminMember } from 'src/app/core/chapters/chapter-admin-member';
import { ChapterAdminMembershipSettings } from 'src/app/core/chapters/chapter-admin-membership-settings';
import { ChapterAdminPaymentSettings } from 'src/app/core/chapters/chapter-admin-payment-settings';
import { ChapterLinks } from 'src/app/core/chapters/chapter-links';
import { ChapterProperty } from 'src/app/core/chapters/chapter-property';
import { ChapterQuestion } from 'src/app/core/chapters/chapter-question';
import { ChapterService } from './chapter.service';
import { ChapterSubscription } from 'src/app/core/chapters/chapter-subscription';
import { ChapterTexts } from 'src/app/core/chapters/chapter-texts';
import { environment } from 'src/environments/environment';
import { HttpStore } from '../http/http-store';
import { HttpUtils } from '../http/http-utils';
import { ServiceResult } from '../service-result';

const baseUrl = `${environment.adminApiBaseUrl}/chapters`;

const endpoints = {
  adminMember: (id: string, memberId: string) => `${baseUrl}/${id}/adminmembers/${memberId}`,
  adminMembers: (id: string) => `${baseUrl}/${id}/adminmembers`,
  chapters: baseUrl,  
  links: (id: string) => `${baseUrl}/${id}/links`,
  membershipSettings: (id: string) => `${baseUrl}/${id}/membership/settings`,
  paymentSettings: (id: string) => `${baseUrl}/${id}/payments/settings`,
  properties: (id: string) => `${baseUrl}/${id}/properties`,
  property: (id: string) => `${baseUrl}/properties/${id}`,
  propertyMoveDown: (id: string) => `${baseUrl}/properties/${id}/movedown`,
  propertyMoveUp: (id: string) => `${baseUrl}/properties/${id}/moveup`,
  question: (id: string) => `${baseUrl}/questions/${id}`,
  questionMoveDown: (id: string) => `${baseUrl}/questions/${id}/movedown`,
  questionMoveUp: (id: string) => `${baseUrl}/questions/${id}/moveup`,
  questions: (id: string) => `${baseUrl}/${id}/questions`,
  subscription: (id: string) => `${baseUrl}/subscriptions/${id}`,
  subscriptions: (id: string) => `${baseUrl}/${id}/subscriptions`,
  texts: (id: string) => `${baseUrl}/${id}/texts`,
};

@Injectable({
  providedIn: 'root'
})
export class ChapterAdminService extends ChapterService {

  constructor(http: HttpClient, store: HttpStore) {
    super(http, store);
  }

  addChapterAdminMember(chapterId: string, memberId: string): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      memberId: memberId
    });

    return this.http.post(endpoints.adminMembers(chapterId), params).pipe(
      map(() => undefined)
    );
  }

  createChapterProperty(chapterId: string, property: ChapterProperty): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      dataType: property.dataType.toString(),
      helpText: property.helpText,
      hidden: property.hidden ? 'True' : 'False',
      label: property.label,
      name: property.name,
      required: property.required ? 'True' : 'False',
      subtitle: property.subtitle
    });

    return this.http.post(endpoints.properties(chapterId), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    )
  }

  createChapterQuestion(chapterId: string, chapterQuestion: ChapterQuestion): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      answer: chapterQuestion.answer,
      name: chapterQuestion.name
    });

    return this.http.post(endpoints.questions(chapterId), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }  

  createChapterSubscription(chapterId: string, subscription: ChapterSubscription): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      amount: subscription.amount.toString(),
      description: subscription.description,
      months: subscription.months.toString(),
      name: subscription.name,
      title: subscription.title,
      type: subscription.type.toString()
    });

    return this.http.post(endpoints.subscriptions(chapterId), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }

  deleteChapterProperty(property: ChapterProperty): Observable<void> {
    return this.http.delete(endpoints.property(property.id)).pipe(
      map(() => undefined)
    );
  }

  deleteChapterQuestion(question: ChapterQuestion): Observable<void> {
    return this.http.delete(endpoints.question(question.id)).pipe(
      map(() => undefined)
    );
  }

  deleteChapterSubscription(subscription: ChapterSubscription): Observable<void> {
    return this.http.delete(endpoints.subscription(subscription.id)).pipe(
      map(() => undefined)
    );
  }

  getAdminChapterProperties(chapterId: string): Observable<ChapterProperty[]> {
    return this.http.get(endpoints.properties(chapterId)).pipe(
      map((response: any) => response.map(x => this.mapChapterProperty(x)))
    );
  }

  getAdminChapters(): Observable<Chapter[]> {
    return this.http.get(endpoints.chapters).pipe(
      map((response: any) => response.map(x => this.mapChapter(x)))
    );
  }  

  getChapterAdminMember(chapterId: string, memberId: string): Observable<ChapterAdminMember> {
    return this.http.get(endpoints.adminMember(chapterId, memberId)).pipe(
      map((response: any) => this.mapChapterAdminMember(response))
    );
  }

  getChapterAdminMembers(chapterId: string): Observable<ChapterAdminMember[]> {
    return this.http.get(endpoints.adminMembers(chapterId)).pipe(
      map((response: any) => response.map(x => this.mapChapterAdminMember(x)))
    );
  }

  getChapterAdminMembershipSettings(chapterId: string): Observable<ChapterAdminMembershipSettings> {
    return this.http.get(endpoints.membershipSettings(chapterId)).pipe(
      map((response: any) => this.mapChapterAdminMembershipSettings(response))
    );
  }

  getChapterAdminPaymentSettings(chapterId: string): Observable<ChapterAdminPaymentSettings> {
    return this.http.get(endpoints.paymentSettings(chapterId)).pipe(
      map((response: any) => this.mapChapterAdminPaymentSettings(response))
    );
  }  

  getChapterProperty(id: string): Observable<ChapterProperty> {
    return this.http.get(endpoints.property(id)).pipe(
      map((response: any) => this.mapChapterProperty(response))
    );
  }

  getChapterQuestion(id: string): Observable<ChapterQuestion> {
    return this.http.get(endpoints.question(id)).pipe(
      map((response: any) => this.mapChapterQuestion(response))
    );
  }

  getChapterSubscription(id: string): Observable<ChapterSubscription> {
    return this.http.get(endpoints.subscription(id)).pipe(
      map((response: any) => this.mapChapterSubscription(response))
    );
  }

  hasAccess(chapter: Chapter): Observable<boolean> {
    return chapter ? this.getAdminChapters().pipe(
      map((chapters: Chapter[]) => !!chapters.find(x => x.id === chapter.id))
    ) : of(false);
  }

  moveChapterPropertyDown(id: string): Observable<ChapterProperty[]> {
    return this.http.put(endpoints.propertyMoveDown(id), {}).pipe(
      map((response: any) => response.map(x => this.mapChapterProperty(x)))
    );
  }

  moveChapterPropertyUp(id: string): Observable<ChapterProperty[]> {
    return this.http.put(endpoints.propertyMoveUp(id), {}).pipe(
      map((response: any) => response.map(x => this.mapChapterProperty(x)))
    );
  }

  moveChapterQuestionDown(id: string): Observable<ChapterQuestion[]> {
    return this.http.put(endpoints.questionMoveDown(id), {}).pipe(
      map((response: any) => response.map(x => this.mapChapterQuestion(x)))
    );
  }

  moveChapterQuestionUp(id: string): Observable<ChapterQuestion[]> {
    return this.http.put(endpoints.questionMoveUp(id), {}).pipe(
      map((response: any) => response.map(x => this.mapChapterQuestion(x)))
    );
  }

  removeChapterAdminMember(chapterId: string, adminMember: ChapterAdminMember): Observable<void> {
    return this.http.delete(endpoints.adminMember(chapterId, adminMember.memberId)).pipe(
      map(() => undefined)
    );
  }

  updateChapterAdminMember(chapterId: string, adminMember: ChapterAdminMember): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      adminEmailAddress: adminMember.adminEmailAddress,
      receiveContactEmails: adminMember.receiveContactEmails ? 'True' : 'False',
      receiveNewMemberEmails: adminMember.receiveNewMemberEmails ? 'True' : 'False',
      sendNewMemberEmails: adminMember.sendNewMemberEmails ? 'True' : 'False'
    });

    return this.http.put(endpoints.adminMember(chapterId, adminMember.memberId), params).pipe(
      map(() => undefined)
    );
  }  

  updateChapterAdminMembershipSettings(chapterId: string, settings: ChapterAdminMembershipSettings): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      membershipDisabledAfterDaysExpired: settings.membershipDisabledAfterDaysExpired.toString(),
      trialPeriodMonths: settings.trialPeriodMonths.toString()
    });

    return this.http.put(endpoints.membershipSettings(chapterId), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }    

  updateChapterAdminPaymentSettings(chapterId: string, settings: ChapterAdminPaymentSettings): Observable<ChapterAdminPaymentSettings> {
    const params: HttpParams = HttpUtils.createFormParams({
      apiPublicKey: settings.apiPublicKey,
      apiSecretKey: settings.apiSecretKey
    });

    return this.http.put(endpoints.paymentSettings(chapterId), params).pipe(
      map((response: any) => this.mapChapterAdminPaymentSettings(response))
    );
  }    

  updateChapterLinks(chapterId: string, links: ChapterLinks): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      facebook: links.facebook,
      instagram: links.instagram,
      twitter: links.twitter
    });

    return this.http.put(endpoints.links(chapterId), params).pipe(
      map(() => undefined)
    );
  }

  updateChapterProperty(property: ChapterProperty): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      helpText: property.helpText,
      hidden: property.hidden ? 'True' : 'False',
      label: property.label,
      name: property.name,
      required: property.required ? 'True' : 'False',
      subtitle: property.subtitle
    });

    return this.http.put(endpoints.property(property.id), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }

  updateChapterQuestion(question: ChapterQuestion): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      answer: question.answer,
      name: question.name
    });

    return this.http.put(endpoints.question(question.id), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }

  updateChapterSubscription(subscription: ChapterSubscription): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      amount: subscription.amount.toString(),
      description: subscription.description,
      months: subscription.months.toString(),
      name: subscription.name,
      title: subscription.title,
      type: subscription.type.toString()
    });

    return this.http.put(endpoints.subscription(subscription.id), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }

  updateChapterTexts(chapterId: string, texts: ChapterTexts): Observable<ChapterTexts> {
    const params: HttpParams = HttpUtils.createFormParams({
      registerText: texts.registerText,
      welcomeText: texts.welcomeText
    });

    return this.http.put(endpoints.texts(chapterId), params).pipe(
      map((response: any) => this.mapChapterTexts(response))
    );
  }

  private mapChapterAdminMember(response: any): ChapterAdminMember {
    return {
      adminEmailAddress: response.adminEmailAddress,
      firstName: response.firstName,
      fullName: `${response.firstName} ${response.lastName}`,
      lastName: response.lastName,
      memberId: response.memberId,
      receiveContactEmails: response.receiveContactEmails,
      receiveNewMemberEmails: response.receiveNewMemberEmails,
      sendNewMemberEmails: response.sendNewMemberEmails
    };
  }

  private mapChapterAdminMembershipSettings(response: any): ChapterAdminMembershipSettings {
    return {
      membershipDisabledAfterDaysExpired: response.membershipDisabledAfterDaysExpired,
      trialPeriodMonths: response.trialPeriodMonths
    };
  }  

  private mapChapterAdminPaymentSettings(response: any): ChapterAdminPaymentSettings {
    return {
      apiPublicKey: response.apiPublicKey,
      apiSecretKey: response.apiSecretKey,
      provider: response.provider
    };
  }  
}

File: ODK.Website.Api/ODK.Services/Chapters/ChapterAdminService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ODK.Core.Chapters;
using ODK.Core.DataTypes;
using ODK.Core.Members;
using ODK.Services.Caching;
using ODK.Services.Exceptions;

namespace ODK.Services.Chapters
{
    public class ChapterAdminService : OdkAdminServiceBase, IChapterAdminService
    {
        private readonly ICacheService _cacheService;
        private readonly IChapterRepository _chapterRepository;
        private readonly IChapterService _chapterService;
        private readonly IMemberRepository _memberRepository;

        public ChapterAdminService(IChapterRepository chapterRepository, ICacheService cacheService,
            IChapterService chapterService, IMemberRepository memberRepository)
            : base(chapterRepository)
        {
            _cacheService = cacheService;
            _chapterRepository = chapterRepository;
            _chapterService = chapterService;
            _memberRepository = memberRepository;
        }

        public async Task AddChapterAdminMember(Guid currentMemberId, Guid chapterId, Guid memberId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            Member member = await _memberRepository.GetMember(memberId);
            if (member == null)
            {
                throw new OdkNotFoundException();
            }

            ChapterAdminMember existing = await _chapterRepository.GetChapterAdminMember(chapterId, memberId);
            if (existing != null)
            {
                return;
            }

            ChapterAdminMember adminMember = new ChapterAdminMember(chapterId, memberId);
            await _chapterRepository.AddChapterAdminMember(adminMember);
        }

        public async Task CreateChapterProperty(Guid currentMemberId, Guid chapterId, CreateChapterProperty property)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            IReadOnlyCollection<ChapterProperty> existing = await _chapterRepository.GetChapterProperties(chapterId);

            int displayOrder = existing.Count > 0 ? existing.Max(x => x.DisplayOrder) + 1 : 1;
            ChapterProperty create = new ChapterProperty(Guid.Empty, chapterId, property.DataType, property.Name?.ToLowerInvariant(),
                property.Label, displayOrder, property.Required, property.Subtitle, property.HelpText, property.Hidden);

            await ValidateChapterProperty(create);

            await _chapterRepository.AddChapterProperty(create);
            _cacheService.RemoveVersionedCollection<ChapterProperty>(chapterId);
        }

        public async Task CreateChapterQuestion(Guid currentMemberId, Guid chapterId, CreateChapterQuestion question)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            IReadOnlyCollection<ChapterQuestion> existing = await _chapterRepository.GetChapterQuestions(chapterId);

            int displayOrder = existing.Count  > 0 ? existing.Max(x => x.DisplayOrder) + 1 : 1;
            ChapterQuestion create = new ChapterQuestion(Guid.Empty, chapterId, question.Name, question.Answer, displayOrder, 0);

            ValidateChapterQuestion(create);

            await _chapterRepository.CreateChapterQuestion(create);

            _cacheService.RemoveVersionedItem<ChapterQuestion>(chapterId);
        }

        public async Task CreateChapterSubscription(Guid currentMemberId, Guid chapterId, CreateChapterSubscription subscription)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterSubscription create = new ChapterSubscription(Guid.Empty, chapterId, subscription.Type, subscription.Name,
                subscription.Title, subscription.Description, subscription.Amount, subscription.Months);

            await ValidateChapterSubscription(create);

            await _chapterRepository.CreateChapterSubscription(create);
        }

        public async Task DeleteChapterAdminMember(Guid currentMemberId, Guid chapterId, Guid memberId)
        {
            ChapterAdminMember adminMember = await GetChapterAdminMember(currentMemberId, chapterId, memberId);
            if (adminMember.SuperAdmin)
            {
                return;
            }

            await _chapterRepository.DeleteChapterAdminMember(chapterId, memberId);
        }

        public async Task DeleteChapterProperty(Guid currentMemberId, Guid id)
        {
            ChapterProperty property = await GetChapterProperty(currentMemberId, id);

            await _chapterRepository.DeleteChapterProperty(id);

            IReadOnlyCollection<ChapterProperty> properties = await _chapterRepository.GetChapterProperties(property.ChapterId, true);
            int displayOrder = 1;
            foreach (ChapterProperty reorder in properties.OrderBy(x => x.DisplayOrder))
            {
                if (reorder.DisplayOrder != displayOrder)
                {
                    reorder.DisplayOrder = displayOrder;
                    await _chapterRepository.UpdateChapterProperty(reorder);
                }

                displayOrder++;
            }

            _cacheService.RemoveVersionedCollection<ChapterProperty>(property.ChapterId);
        }

        public async Task DeleteChapterQuestion(Guid currentMemberId, Guid id)
        {
            ChapterQuestion question = await GetChapterQuestion(currentMemberId, id);

            await _chapterRepository.DeleteChapterQuestion(id);

            IReadOnlyCollection<ChapterQuestion> questions = await _chapterRepository.GetChapterQuestions(question.ChapterId);
            int displayOrder = 1;
            foreach (ChapterQuestion reorder in questions.OrderBy(x => x.DisplayOrder))
            {
                if (reorder.DisplayOrder != displayOrder)
                {
                    reorder.DisplayOrder = displayOrder;
                    await _chapterRepository.UpdateChapterQuestion(reorder);
                }

                displayOrder++;
            }

            _cacheService.RemoveVersionedCollection<ChapterQuestion>(question.ChapterId);
        }

        public async Task DeleteChapterSubscription(Guid currentMemberId, Guid id)
        {
            ChapterSubscription subscription = await GetChapterSubscription(currentMemberId, id);

            await _chapterRepository.DeleteChapterSubscription(subscription.Id);
        }

        public async Task<ChapterAdminMember> GetChapterAdminMember(Guid currentMemberId, Guid chapterId, Guid memberId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterAdminMember adminMember = await _chapterRepository.GetChapterAdminMember(chapterId, memberId);
            if (adminMember == null)
            {
                throw new OdkNotFoundException();
            }

            return adminMember;
        }

        public async Task<IReadOnlyCollection<ChapterAdminMember>> GetChapterAdminMembers(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterAdminMembers(chapterId);
        }

        public async Task<ChapterMembershipSettings> GetChapterMembershipSettings(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterMembershipSettings(chapterId);
        }

        public async Task<ChapterPaymentSettings> GetChapterPaymentSettings(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterSuperAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterPaymentSettings(chapterId);
        }

        public async Task<IReadOnlyCollection<ChapterProperty>> GetChapterProperties(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterProperties(chapterId, true);
        }

        public async Task<ChapterProperty> GetChapterProperty(Guid currentMemberId, Guid id)
        {
            ChapterProperty property = await _chapterRepository.GetChapterProperty(id);
            if (property == null)
            {
                throw new OdkNotFoundException();
            }

            await AssertMemberIsChapterAdmin(currentMemberId, property.ChapterId);

            return property;
        }

        public async Task<ChapterQuestion> GetChapterQuestion(Guid currentMemberId, Guid questionId)
        {
            ChapterQuestion question = await _chapterRepository.GetChapterQuestion(questionId);
            if (question == null)
            {
                throw new OdkNotFoundException();
            }

            await AssertMemberIsChapterAdmin(currentMemberId, question.ChapterId);

            return question;
        }

        public async Task<IReadOnlyCollection<Chapter>> GetChapters(Guid currentMemberId)
        {
            IReadOnlyCollection<ChapterAdminMember> chapterAdminMembers = await _chapterRepository.GetChapterAdminMembersByMember(currentMemberId);
            if (chapterAdminMembers.Count == 0)
            {
                throw new OdkNotAuthorizedException();
            }
            VersionedServiceResult<IReadOnlyCollection<Chapter>> chapters = await _chapterService.GetChapters(null);
            return chapterAdminMembers
                .Select(x => chapters.Value.Single(chapter => chapter.Id == x.ChapterId))
                .ToArray();
        }

        public async Task<ChapterSubscription> GetChapterSubscription(Guid currentMemberId, Guid id)
        {
            ChapterSubscription subscription = await _chapterRepository.GetChapterSubscription(id);
            if (subscription == null)
            {
                throw new OdkNotFoundException();
            }

            await AssertMemberIsChapterAdmin(currentMemberId, subscription.ChapterId);

            return subscription;
        }

        public async Task<IReadOnlyCollection<ChapterSubscription>> GetChapterSubscriptions(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterSubscriptions(chapterId);
        }

        public async Task UpdateChapterAdminMember(Guid currentMemberId, Guid chapterId, Guid memberId,
            UpdateChapterAdminMember adminMember)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterAdminMember existing = await _chapterRepository.GetChapterAdminMember(chapterId, memberId);
            if (existing == null)
            {
                throw new OdkNotFoundException();
            }

            existing.AdminEmailAddress = adminMember.AdminEmailAddress;
            existing.ReceiveContactEmails = adminMember.ReceiveContactEmails;
            existing.ReceiveNewMemberEmails = adminMember.ReceiveNewMemberEmails;
            existing.SendNewMemberEmails = adminMember.SendNewMemberEmails;

            await _chapterRepository.UpdateChapterAdminMember(existing);
        }

        public async Task UpdateChapterLinks(Guid currentMemberId, Guid chapterId, UpdateChapterLinks links)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterLinks update = new ChapterLinks(chapterId, links.Facebook, links.Instagram, links.Twitter, 0);
            await _chapterRepository.UpdateChapterLinks(update);

            _cacheService.RemoveVersionedItem<ChapterLinks>(chapterId);
        }

        public async Task UpdateChapterMembershipSettings(Guid currentMemberId, Guid chapterId, UpdateChapterMembershipSettings settings)
        {
            ChapterMembershipSettings update = await GetChapterMembershipSettings(currentMemberId, chapterId);

            update.MembershipDisabledAfterDaysExpired = settings.MembershipDisabledAfterDaysExpired;
            update.TrialPeriodMonths = settings.TrialPeriodMonths;

            ValidateChapterMembershipSettings(update);

            await _chapterRepository.UpdateChapterMembershipSettings(update);

            _cacheService.UpdateItem(update, chapterId);
        }

        public async Task<ChapterPaymentSettings> UpdateChapterPaymentSettings(Guid currentMemberId, Guid chapterId,
            UpdateChapterPaymentSettings settings)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterPaymentSettings existing = await _chapterRepository.GetChapterPaymentSettings(chapterId);
            ChapterPaymentSettings update = new ChapterPaymentSettings(chapterId, settings.ApiPublicKey, settings.ApiSecretKey, existing.Provider);

            await _chapterRepository.UpdateChapterPaymentSettings(update);

            return update;
        }

        public async Task UpdateChapterProperty(Guid currentMemberId, Guid propertyId, UpdateChapterProperty property)
        {
            ChapterProperty update = await GetChapterProperty(currentMemberId, propertyId);

            update.HelpText = property.HelpText;
            update.Hidden = property.Hidden;
            update.Label = property.Label;
            update.Name = property.Name?.ToLowerInvariant();
            update.Required = property.Required;
            update.Subtitle = property.Subtitle;

            await ValidateChapterProperty(update);

            await _chapterRepository.UpdateChapterProperty(update);

            _cacheService.RemoveVersionedCollection<ChapterProperty>(update.ChapterId);
        }

        public async Task<IReadOnlyCollection<ChapterProperty>> UpdateChapterPropertyDisplayOrder(Guid currentMemberId, Guid propertyId, int moveBy)
        {
            ChapterProperty property = await GetChapterProperty(currentMemberId, propertyId);
            IReadOnlyCollection<ChapterProperty> properties = await _chapterRepository.GetChapterProperties(property.ChapterId);

            if (moveBy == 0)
            {
                return properties;
            }

            ChapterProperty switchWith;
            if (moveBy > 0)
            {
                switchWith = properties
                    .Where(x => x.DisplayOrder > property.DisplayOrder)
                    .OrderBy(x => x.DisplayOrder)
                    .FirstOrDefault();

            }
            else
            {
                switchWith = properties
                    .Where(x => x.DisplayOrder < property.DisplayOrder)
                    .OrderByDescending(x => x.DisplayOrder)
                    .FirstOrDefault();
            }

            if (switchWith == null)
            {
                return properties;
            }

            property = properties.First(x => x.Id == property.Id);

            int displayOrder = switchWith.DisplayOrder;
            switchWith.DisplayOrder = property.DisplayOrder;
            property.DisplayOrder = displayOrder;

            await _chapterRepository.UpdateChapterProperty(property);
            await _chapterRepository.UpdateChapterProperty(switchWith);

            _cacheService.RemoveVersionedCollection<ChapterProperty>(property.ChapterId);

            return properties.OrderBy(x => x.DisplayOrder).ToArray();
        }

        public async Task UpdateChapterQuestion(Guid currentMemberId, Guid questionId, CreateChapterQuestion question)
        {
            ChapterQuestion update = await GetChapterQuestion(currentMemberId, questionId);

            update.Answer = question.Answer;
            update.Name = question.Name;

            ValidateChapterQuestion(update);

            await _chapterRepository.UpdateChapterQuestion(update);

            _cacheService.RemoveVersionedCollection<ChapterQuestion>(update.ChapterId);
        }

        public async Task<IReadOnlyCollection<ChapterQuestion>> UpdateChapterQuestionDisplayOrder(Guid currentMemberId, Guid questionId, int moveBy)
        {
            ChapterQuestion question = await GetChapterQuestion(currentMemberId, questionId);
            IReadOnlyCollection<ChapterQuestion> questions = await _chapterRepository.GetChapterQuestions(question.ChapterId);

            if (moveBy == 0)
            {
                return questions;
            }

            ChapterQuestion switchWith;
            if (moveBy > 0)
            {
                switchWith = questions
                    .Where(x => x.DisplayOrder > question.DisplayOrder)
                    .OrderBy(x => x.DisplayOrder)
                    .FirstOrDefault();

            }
            else
            {
                switchWith = questions
                    .Where(x => x.DisplayOrder < question.DisplayOrder)
                    .OrderByDescending(x => x.DisplayOrder)
                    .FirstOrDefault();
            }

            if (switchWith == null)
            {
                return questions;
            }

            question = questions.First(x => x.Id == question.Id);

            int displayOrder = switchWith.DisplayOrder;
            switchWith.DisplayOrder = question.DisplayOrder;
            question.DisplayOrder = displayOrder;

            await _chapterRepository.UpdateChapterQuestion(question);
            await _chapterRepository.UpdateChapterQuestion(switchWith);

            _cacheService.RemoveVersionedCollection<ChapterQuestion>(question.ChapterId);

            return questions.OrderBy(x => x.DisplayOrder).ToArray();
        }

        public async Task UpdateChapterSubscription(Guid currentMemberId, Guid id, CreateChapterSubscription subscription)
        {
            ChapterSubscription existing = await _chapterRepository.GetChapterSubscription(id);
            if (existing == null)
            {
                throw new OdkNotFoundException();
            }

            await AssertMemberIsChapterAdmin(currentMemberId, existing.ChapterId);

            existing.Amount = subscription.Amount;
            existing.Description = subscription.Description;
            existing.Months = subscription.Months;
            existing.Name = subscription.Name;
            existing.Title = subscription.Title;
            existing.Type = subscription.Type;

            await ValidateChapterSubscription(existing);

            await _chapterRepository.UpdateChapterSubscription(existing);
        }

        public async Task<ChapterTexts> UpdateChapterTexts(Guid currentMemberId, Guid chapterId, UpdateChapterTexts texts)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            if (string.IsNullOrWhiteSpace(texts.RegisterText) ||
                string.IsNullOrWhiteSpace(texts.WelcomeText))
            {
                throw new OdkServiceException("Some required fields are missing");
            }

            ChapterTexts update = new ChapterTexts(chapterId, texts.RegisterText, texts.WelcomeText);

            await _chapterRepository.UpdateChapterTexts(update);

            _cacheService.RemoveVersionedItem<ChapterTexts>(chapterId);

            return update;
        }

        private void ValidateChapterQuestion(ChapterQuestion question)
        {
            if (string.IsNullOrWhiteSpace(question.Name) ||
                string.IsNullOrWhiteSpace(question.Answer))
            {
                throw new OdkServiceException("Some required fields are missing");
            }
        }

        private void ValidateChapterMembershipSettings(ChapterMembershipSettings settings)
        {
            if (settings.MembershipDisabledAfterDaysExpired < 0 ||
                settings.TrialPeriodMonths <= 0)
            {
                throw new OdkServiceException("Some required fields are missing");
            }
        }

        private async Task ValidateChapterProperty(ChapterProperty property)
        {
            if (string.IsNullOrEmpty(property.Name) ||
                string.IsNullOrEmpty(property.Label) ||
                !Enum.IsDefined(typeof(DataType), property.DataType) || property.DataType == DataType.None)
            {
                throw new OdkServiceException("Some required fields are missing");
            }

            IReadOnlyCollection<ChapterProperty> properties = await _chapterRepository.GetChapterProperties(property.ChapterId);
            if (properties.Any(x => x.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase) && x.Id != property.Id))
            {
                throw new OdkServiceException("Name already exists");
            }
        }

        private async Task ValidateChapterSubscription(ChapterSubscription subscription)
        {
            if (!Enum.IsDefined(typeof(SubscriptionType), subscription.Type) || subscription.Type == SubscriptionType.None)
            {
                throw new OdkServiceException("Invalid type");
            }

            if (string.IsNullOrWhiteSpace(subscription.Description) ||
                string.IsNullOrWhiteSpace(subscription.Name) ||
                string.IsNullOrWhiteSpace(subscription.Title))
            {
                throw new OdkServiceException("Some required fields are missing");
            }

            if (subscription.Amount < 0)
            {
                throw new OdkServiceException("Amount cannot be less than 0");
            }

            if (subscription.Months < 1)
            {
                throw new OdkServiceException("Subscription must be for at least 1 month");
            }

            IReadOnlyCollection<ChapterSubscription> existing = await _chapterRepository.GetChapterSubscriptions(subscription.ChapterId);
            if (existing.Any(x => x.Id != subscription.Id && x.Name.Equals(subscription.Name)))
            {
                throw new OdkServiceException("A subscription with that name already exists");
            }
        }
    }
}
